### PR TITLE
AvA support for distributed applications

### PR DESCRIPTION
**Pull request by yuhc**
This PR introduces support for distributed apps such as `lonestar/analytics/distributed/pagerank` and `bfs`.

An unfixed issue is that libava only supports the non-distributed GPU applications at commit 5c97d652ccfa0e7880e42797fd2615ec17f66644 (dist apps at the latest commit work). Commit 769d93219add7b18317f5904303956c2018a6462 changed the GPU apps' build behavior and some CUDA internel APIs like `__cudaRegisterFatBinary` can't be hooked by libava. I guess that a static CUDA library is linked into the binary.
I spent a few days on checking those commits, but haven't figured out the reason or the way to fix. It may be supposed to be done next.

### CODE REVIEW

File: libava/specs/cuda_10_1.c [hunk 1]
```diff
@@ -159,7 +191,8 @@ ava_utility void __helper_dump_fatbin(void *fatCubin,
     }
 
     /*  Open the command pipe for reading */
-    fp_pipe = popen("/usr/local/cuda-10.1/bin/cuobjdump -elf /tmp/fatbin.cubin", "r");
+    sprintf(cuobjdump_cmd, "/usr/local/cuda-10.1/bin/cuobjdump -elf %s", fatbin_path);
```
Comment (ddn0) [inline libava/specs/cuda_10_1.c:199]: I know that the static sizes are probably adequate, but `snprintf` would be a little more defensive in the face of future changes.

File: libava/manager/superbuild/CMakeLists.txt [hunk 2]
```diff
@@ -45,6 +45,7 @@ ExternalProject_Add(protobuf
         -Dprotobuf_BUILD_TESTS:BOOL=OFF
         -Dprotobuf_WITH_ZLIB:BOOL=OFF
         -Dprotobuf_MSVC_STATIC_RUNTIME:BOOL=OFF
+        -DCMAKE_POSITION_INDEPENDENT_CODE:BOOL=ON
```
Comment (ddn0) [inline libava/manager/superbuild/CMakeLists.txt:48]: qq: why is this required now? and would `BUILD_SHARED_LIBS=on` [1] more directly express what you need?

[1] https://cmake.org/cmake/help/v3.0/variable/BUILD_SHARED_LIBS.html

File: libava/manager/manager.cpp [hunk 3]
```diff
@@ -374,7 +373,7 @@ class ManagerServiceImpl final : public ManagerService::Service {
     /* Look up available GPUs on every node. */
     std::vector<std::shared_ptr<GpuListEntry>> assigned_entries;
     unsigned daemon_idx = 0;
-    for (; config->daemons_.size(); ++daemon_idx) {
+    for (; daemon_idx < config->daemons_.size(); ++daemon_idx) {
```
Comment (ddn0) [inline libava/manager/manager.cpp:389]: nice.

Perhaps `for (unsigned daemon_idx = 0; .......)` now?

File: libava/specs/cuda_10_1.c [hunk 4]
```diff
@@ -160,19 +161,23 @@ ava_utility void __helper_dump_fatbin(void *fatCubin,
...
     close(fd);
     srand(rand_seed);
     __helper_rand_str(random_str, 16);
-    //sprintf(fatbin_path, "/tmp/fatbin_%s.cubin", random_str);
+    sprintf(fatbin_path, "/tmp/fatbin_%s.cubin", random_str);
 
     /* Dump fat binary to a file */
     fd = open(fatbin_path, O_WRONLY | O_TRUNC | O_CREAT, 0666);
-    write(fd, (const void *)wp->ptr, fbh->headerSize + fbh->fatSize);
+    rw_ret = write(fd, (const void *)wp->ptr, fbh->headerSize + fbh->fatSize);
```
Comment (ddn0) [inline libava/specs/cuda_10_1.c:176]: maybe check for `errno` too?

File: libava/manager/manager.cpp [hunk 5]
```diff
@@ -179,11 +179,8 @@ class ManagerServiceImpl final : public ManagerService::Service {
       const flatbuffers::grpc::Message<DaemonRegisterRequest>* request_msg,
       flatbuffers::grpc::Message<DaemonRegisterReply>* response_msg) override {
     const DaemonRegisterRequest* request = request_msg->GetRoot();
-    std::string peer_address =
-        context->peer().substr(context->peer().find(':') + 1);
-    std::string daemon_ip = peer_address.substr(0, peer_address.find(':'));
-    const std::string daemon_address =
-        daemon_ip + ":" + request->daemon_address()->str();
+    const std::string daemon_address     = request->daemon_address()->str();
```
Comment (ddn0) [inline libava/manager/manager.cpp:182]: for later: but it might be nice to have a specific address type to bundle together accessing the port and IP and distinguish std::string from addresses

File: libava/manager/manager.h [hunk 6]
```diff
@@ -195,6 +195,8 @@ class GpuList {
                   return a->GetFreeMemory() > b->GetFreeMemory();
                 return a->worker_counter_ < b->worker_counter_;
               });
+    std::cout << gpu_list_[0]->GetUuid() << ", "
+              << gpu_list_[0]->worker_counter_ << std::endl;
```
Comment (ddn0) [inline libava/manager/manager.h:199]: nit: `std::endl` => `\n`

File: libava/specs/cuda_10_1.c [hunk 7]
```diff
@@ -132,15 +132,46 @@ ava_utility int __helper_cubin_num(void **cubin_handle) {
...
+    int64_t rand_seed;
+    ssize_t rand_ret = read(fd, &rand_seed, sizeof(rand_seed));
+    if (rand_ret <= 0) {
+      fprintf(stderr, "Failed read from /dev/urandom\n");
+      exit(1);
+    }
+    close(fd);
+    srand(rand_seed);
+    __helper_rand_str(random_str, 16);
+    //sprintf(fatbin_path, "/tmp/fatbin_%s.cubin", random_str);
```
[the remark this answers is not in the document]
Comment (yuhc) [reply libava/specs/cuda_10_1.c]: It was commented out by mistake and I found `fatbin_path` empty when I was testing the next commit.
I'll fix it later.

File: libava/manager/superbuild/CMakeLists.txt (hunk 2 above)
Comment (yuhc) [reply libava/manager/superbuild/CMakeLists.txt:48]: I believe `BUILD_SHARED_LIBS=on` indicates `POSITION_INDEPENDENT_CODE=on`. I'll remove this line.

File: libava/manager/manager.h (hunk 6 above)
Comment (yuhc) [reply libava/manager/manager.h:199]: This was actually a debug message :) . Forgot to remove it.

File: libava/manager/superbuild/CMakeLists.txt (hunk 2 above)
Comment (yuhc) [reply libava/manager/superbuild/CMakeLists.txt:48]: Oh `POSITION_INDEPENDENT_CODE=on` is preferred because I want `protobuf` to be linked to `grpc` statically.
Then `grpc` is linked to manager and spawn daemon statically (`libgrpc.a`); but it should be linked to `guestlib` dynamically (`libgrpc.so`) (otherwise `grpc` APIs don't work in the guestlib's constructor).

Because we don't install the built protobuf or grpc into the system paths, using static libraries would minimize the changes to the library search path.
By default, `protobuf` has `POSITION_INDEPENDENT_CODE=on` but I put in this line for safety.

I think it's a good time to use conan to manage these grpc dependencies and get rid of these external projects.

File: libava/specs/cuda_10_1.c (hunk 4 above)
Comment (yuhc) [reply libava/specs/cuda_10_1.c:176]: At error, rw_ret will be -1. It's safe without checking errno. (I changed the fprintf below to print out `errno`).

File: libava/manager/manager.cpp (hunk 5 above)
Comment (yuhc) [reply libava/manager/manager.cpp:182]: That'll be very clean--I thought about it and will do in next PR.

File: libava/specs/cuda_10_1.c [hunk 8]
```diff
@@ -159,7 +197,8 @@ ava_utility void __helper_dump_fatbin(void *fatCubin,
     }
 
     /*  Open the command pipe for reading */
-    fp_pipe = popen("/usr/local/cuda-10.1/bin/cuobjdump -elf /tmp/fatbin.cubin", "r");
+    snprintf(cuobjdump_cmd, 75, "/usr/local/cuda-10.1/bin/cuobjdump -elf %s", fatbin_path);
+    fp_pipe = popen(cuobjdump_cmd, "r");
     assert(fp_pipe);
```
Comment (tylershunt) [inline libava/specs/cuda_10_1.c:202]: This `assert` will get compiled out in a release build right? Might handle an error here like you do other places in this file.

Comment (yuhc) [reply libava/specs/cuda_10_1.c]: Fixed. (It's a bad coding style. I also need to fix it in lots of ava codes :( )
